Account app updates and cleanups

kschiffer opened the pull request:
#### Summary
This PR introduces fixes and updates to the current scaffold, which will prepare it for the upcoming PR on profile settings. It also contains refactors that will conform it to our current code style-guide, including the exclusive use of functional components, usage of arrow components. As such, the goal is also to reduce the diff of upcoming PRs.

#### Changes
- Convert all functions to arrow functions #3540 
- Convert all class components to functional components with hooks
- Remove context based usage of `env` (via `withEnv()` and `<EnvProvider />`)
- Update change password functionality
  - Use container class (to be used later-on in the authenticated view as well)
  - Use password validation rules from the IS (see #1067)
  - Add `updateUserPassword` rpc to JS SDK
- Add `requestPromiseMiddleware`
- Add redux primitives and logic for IS config
- Add SDK instance singleton module
- Some logic updates (using `createRequestLogic`)
- Optimize update password flow (e.g. redirect to login automatically)
- Fix routing
  - Handle root route via frontend (instead of backend redirect)
  - Use `?n=…` query parameter only for known authenticated routes 
  - Use not found page otherwise
- Replace usage of API tree with direct usage of the SDK instance
- Various small related fixes

#### Testing

I'm currently working on #3605 to be able to add some more tests, but would like this to be reviewed in the meantime already.

##### Regressions

All original OAuth functionality might experience regression with this.

#### Notes for Reviewers
Sorry for the big diff, it is mainly due to the functional component and arrow function updates. This could have been done successively as well but I really want the AA to be proper right from the start.

Again note that this PR targets the `feature/1422-account-app` feature branch, which I kept rebased on latest master.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

kschiffer committed:
oauth: Add request promise middleware

kschiffer committed:
util: Add change password rpc to user service

kschiffer committed:
account: Convert header container to functional component

kschiffer committed:
account: Apply password validation rules to change password form

kschiffer committed:
account: Add is config redux primitives

kschiffer committed:
account: Fix root routing

kschiffer committed:
account: Fix create account form

kschiffer committed:
account: Fix login form

kschiffer committed:
account: Use new change password form

kschiffer committed:
console,account: Add lib to create password validation schemas

kschiffer committed:
account: Fix state update on unmounted component

kschiffer committed:
dev: Update locales

kschiffer committed:
account: Use request logic util for user logic

kschiffer committed:
account: Confrom entrypoint to latest styleguide

kschiffer committed:
account: Convert to arrow functions

kschiffer committed:
account: Convert views to functional components

kschiffer committed:
console,account: Remove env prop requirement from erro view util

kschiffer committed:
console,account: Remove env prop requirement from locale provider

kschiffer committed:
account: Handle root route redirect in frontend code

kschiffer committed:
account: Optimize forgot password flow

kschiffer commented on pkg/webui/account/views/overview/index.js:
```diff
@@ -0,0 +1,68 @@
+    logout,
+  },
+)
+export default class Overview extends React.PureComponent {
```
Note that this will be completely replaced in the next PR, which I why I did not update this one to a functional component.

kschiffer committed:
dev: Update cypress tests

kschiffer committed:
account: Add CSRF retry interceptor to axios instance

kschiffer committed:
console,account: Extend error http status utility for axios responses

kschiffer committed:
account: Convert api tree to arrow functions

kschiffer committed:
account: Use SDK instance for user creation

kschiffer committed:
util: Add invitation token parameter to user create function

kschiffer committed:
account: Replace temporary password API with SDK instance

kschiffer committed:
util: Add temporary password RPC to JS SDK

kschiffer committed:
account: Remove password update from api tree

kschiffer committed:
account: Use SDK instance for contact info validation

kschiffer committed:
util: Add contact info service to JS SDK

kschiffer committed:
account: Add sdk client singleton module

bafonins commented on pkg/webui/account/views/create-account/index.js:
```diff
@@ -91,151 +86,107 @@ const getSuccessMessage = state => {
+          onSubmit={handleSubmit}
+          initialValues={initialValues}
+          error={error}
+          validationSchema={validationSchema}
```
Any reasons for not using https://github.com/TheThingsNetwork/lorawan-stack/blob/feature/1422-account-app-cleanups/pkg/webui/lib/create-password-validation-schema.js for this form?

bafonins commented on pkg/webui/account/views/forgot-password/index.js:
```diff
@@ -58,81 +55,66 @@ const initialValues = { user_id: '' }
+          component={SubmitButton}
+          message={m.send}
+          className={style.submitButton}
+          alwaysEnabled
```
Cant find the `alwaysEnabled` prop defined anywhere

bafonins commented on pkg/webui/account/views/forgot-password/index.js:
```diff
@@ -58,81 +55,66 @@ const initialValues = { user_id: '' }
+          className={style.submitButton}
+          alwaysEnabled
+        />
+        <Button.Link naked secondary message={sharedMessages.cancel} to="/login" />
```
1. Navigate to `/console`
2. Land on the login page and go to the forgot password view
3. Submit or just navigate back
4. Login
5. Land on the account app overview instead of the Console

Not sure if we should preserve the initial users location, but is seems weird that we do that for registration and not for the forgot password view.

bafonins commented on pkg/webui/account/views/front/index.js:
```diff
@@ -43,6 +47,11 @@ const FrontView = () => {
             <Redirect exact from="/" to="/login" />
+            {authRoutes.map(({ path, exact }) => (
+              <Route path={path} exact={exact} key={path}>
+                <Redirect to={`/login?n=${selectApplicationRootPath()}${location.pathname}`} />
```
What is the purpose of this redirect?

kschiffer replied:
The purpose is to redirect routes that require authentication back to the login page, with a query parameter that redirectes to the original route after login. E.g. `/profile-settings` (will be added later) would redirect back to the login page. All other routes that do not correspond to anything will show a 404.

Note that the `<Front />` handler only operates when the user is not authenticated.

bafonins commented on pkg/webui/account/views/front/index.js:
```diff
@@ -43,6 +47,11 @@ const FrontView = () => {
             <Route path="/validate" component={Validate} />
             <Route path="/code" component={Code} />
             <Redirect exact from="/" to="/login" />
+            {authRoutes.map(({ path, exact }) => (
```
Currently, there is only one route in `authRouter` which is for `/`. When the user navigates to `/oauth` these lines are never reached because of L49

kschiffer replied:
That's true. There will be more routes added with the next PRs which will make this setup purposeful. It's fine that `/` is handled by L49, since we don't need a redirect query parameter for it.

bafonins commented on pkg/webui/account/views/front/index.js:
```diff
@@ -26,10 +26,14 @@ import UpdatePassword from '@account/views/update-password'
 import FrontNotFound from '@account/views/front-not-found'
 import Validate from '@account/views/validate'
 import Code from '@account/views/code'
+import { authRoutes } from '@account/views/landing'
```
IMO, moving this to a separate file is better.

bafonins reviewed (commented):
Submit section on forms looks a bit weird on tablets:

<img width="579" alt="Screenshot 2020-12-27 at 16 33 35" src="https://user-images.githubusercontent.com/16374166/103173654-00f6c080-4865-11eb-83cb-f09ddfd48718.png">


Page content on the account overview completely disappears on phones:

<img width="413" alt="Screenshot 2020-12-27 at 16 35 39" src="https://user-images.githubusercontent.com/16374166/103173669-1835ae00-4865-11eb-8da5-7c9b0f138bf3.png">

kschiffer committed:
account: Use dynamic password validation in registration form

kschiffer committed:
account: Retain query parameter during forgot password flow

kschiffer committed:
account: Improve front mobile styles

kschiffer committed:
account: Use constant module for authenticated paths

kschiffer commented:
Thanks for the review. I've addressed all your comments.

bafonins commented on pkg/webui/lib/create-password-validation-schema.js:
```diff
@@ -0,0 +1,56 @@
+    )
+    .test(
+      'min-special',
+      { message: sharedMessages.validateSpecial, values: { special: requirements.min_special } },
```
`validateSpecial` is not defined in `shared-messages`

bafonins commented on pkg/webui/lib/create-password-validation-schema.js:
```diff
@@ -0,0 +1,56 @@
+    )
+    .test(
+      'min-upper',
+      { message: sharedMessages.validateUppercase, values: { upper: requirements.min_uppercase } },
```
`validateUppercase` is not defined in `shared-messages`

bafonins commented on pkg/webui/lib/create-password-validation-schema.js:
```diff
@@ -0,0 +1,56 @@
+    )
+    .test(
+      'min-digit',
+      { message: sharedMessages.validateDigit, values: { digit: requirements.min_digits } },
```
`validateDigit` is not defined in `shared-messages`

bafonins commented on pkg/webui/account/views/validate/index.js:
```diff
@@ -45,70 +46,74 @@ const m = defineMessages({
       }
     }
-  }
+    makeRequest()
```
Any reason for extracting the whole logic to a local function?

kschiffer replied:
Because we need to run async logic there and it's not allowed to use an async function directly as argument for `useEffect()`, since the effect function must either return a cleanup function or nothing (but would then return a promise).

kschiffer committed:
account: Use store as singleton module

kschiffer committed:
account: Streamline user action names

kschiffer committed:
account: Update initialization logic

bafonins commented on pkg/webui/account/store/middleware/user.js:
```diff
@@ -12,28 +12,38 @@
+    const {
+      meta: { selector },
+    } = action
+    console.log(action)
```
Remove

bafonins commented:
1. Go to `/oauth`
2. Fill in login/password and submit the form
3. Results in an endless loop without being logged in to the account app

Configuration I used for testing:

```
go run ./cmd/ttn-lw-stack start all \
--log.level debug \
--is.oauth.ui.js-file libs.bundle.js,account.js \
--console.ui.js-file libs.bundle.js,console.js \
--console.ui.assets-base-url http://localhost:8080/assets \
--console.ui.canonical-url http://localhost:8080/console \
--console.ui.is.enabled --console.ui.is.base-url http://localhost:8080/api/v3 \
--console.ui.as.enabled --console.ui.as.base-url http://localhost:8080/api/v3 \
--console.ui.ns.enabled --console.ui.ns.base-url http://localhost:8080/api/v3 \
--console.ui.js.enabled --console.ui.js.base-url http://localhost:8080/api/v3 \
--console.ui.gs.enabled --console.ui.gs.base-url http://localhost:8080/api/v3 \
--console.oauth.logout-url=http://localhost:8080/oauth/logout \
--console.oauth.authorize-url=http://localhost:8080/oauth/authorize \
--console.oauth.token-url=http://localhost:8080/oauth/token
```

kschiffer commented:
Hmm that's odd. I cannot reproduce this. Does the same thing happen in incognito mode? It also doesn't seem to be reflected by the E2Es.

bafonins commented:
> Does the same thing happen in incognito mode?

yes, same thing.

> It also doesn't seem to be reflected by the E2Es.

Try adding `cy.wait(10000)` at the end of `succeeds logging in when using valid credentials` test. Same issue happens when running cypress.

I guess we should revisit the way we assert page load

kschiffer commented:
Got it now, you're missing `--is.oauth.ui.is.base-url http://localhost:8080/api/v3` in your configuration. Since this is a different host, the session cookie will not be sent to the IS making it return a 401.

Try like this:

```shell
go run ./cmd/ttn-lw-stack start \
--log.level debug \
--is.oauth.ui.js-file libs.bundle.js,account.js \
--console.ui.js-file libs.bundle.js,console.js \
--console.ui.assets-base-url http://localhost:8080/assets \
--console.ui.canonical-url http://localhost:8080/console \
--is.oauth.ui.canonical-url http://localhost:8080/oauth \
--is.oauth.ui.is.enabled --is.oauth.ui.is.base-url http://localhost:8080/api/v3 \
--console.ui.is.enabled --console.ui.is.base-url http://localhost:8080/api/v3 \
--console.ui.as.enabled --console.ui.as.base-url http://localhost:8080/api/v3 \
--console.ui.ns.enabled --console.ui.ns.base-url http://localhost:8080/api/v3 \
--console.ui.js.enabled --console.ui.js.base-url http://localhost:8080/api/v3 \
--console.ui.gs.enabled --console.ui.gs.base-url http://localhost:8080/api/v3 \
--console.oauth.logout-url=http://localhost:8080/oauth/logout \
--console.oauth.authorize-url=http://localhost:8080/oauth/authorize \
--console.oauth.token-url=http://localhost:8080/oauth/token
```

These are the additional ENVs I use on top [of the usual dev config setup](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.10/DEVELOPMENT.md#development-configuration):
```
export TTN_LW_IS_OAUTH_UI_JS_FILE="libs.bundle.js account.js"
export TTN_LW_IS_OAUTH_UI_IS_BASE_URL="http://localhost:8080/api/v3"
```

htdvisser commented on cmd/internal/shared/identityserver/config.go:
```diff
@@ -28,15 +28,17 @@ import (
 				CanonicalURL:  shared.DefaultOAuthPublicURL,
 				AssetsBaseURL: shared.DefaultAssetsBaseURL,
 				IconPrefix:    "oauth-",
-				CSSFiles:      []string{"oauth.css"},
-				JSFiles:       []string{"oauth.js"},
+				CSSFiles:      []string{"account.css"},
+				JSFiles:       []string{"account.js"},
```
This is a breaking change for deployments that use custom JS/CSS files (which I don't think is very common). In any case, make sure to mention in CHANGELOG that deployments need to change their config if they customized these.

htdvisser commented on pkg/identityserver/identityserver.go:
```diff
@@ -106,10 +108,18 @@ func New(c *component.Component, config *Config) (is *IdentityServer, err error)
 		store.ClientStore
 		store.OAuthStore
 	}{
-		UserStore:        store.GetUserStore(is.db),
-		UserSessionStore: store.GetUserSessionStore(is.db),
 		ClientStore:      store.GetClientStore(is.db),
 		OAuthStore:       store.GetOAuthStore(is.db),
+		UserStore:        store.GetUserStore(is.db),
+		UserSessionStore: store.GetUserSessionStore(is.db),
```
Unnecessary diff

htdvisser replied:
I'm still seeing the diff in the latest version of the PR

htdvisser reviewed (commented):
Please update CODEOWNERS so that there are code owners on `pkg/account`

kschiffer committed:
dev: Update changelog

kschiffer committed:
account: Update logo

htdvisser reviewed (changes requested):
This still doesn't have CODEOWNERS on `pkg/account` as I requested in an earlier review. Please address previous comments before re-requesting review.

kschiffer committed:
dev: Update CODEOWNERS for account package

kschiffer commented:
> This still doesn't have CODEOWNERS on `pkg/account` as I requested in an earlier review. Please address previous comments before re-requesting review.

Sorry, missed a push there.

johanstokking commented on CODEOWNERS:
```diff
@@ -58,6 +58,7 @@
 /pkg/identityserver @htdvisser @bafonins @kschiffer
 /pkg/email @htdvisser @bafonins @kschiffer
 /pkg/oauth @htdvisser @bafonins @kschiffer
+/pkg/account @htdvisser @bafonins @kschiffer
```
This is my only code owned file and I'm fine with this if this was discussed before.

Maybe loop in @asmulko or @mjamescompton instead of @htdvisser to reduce his scope

htdvisser replied:
We should indeed take a better look at the packages that are code-owned by me. In this case it's fine to add me me. (1) It's pretty much all code that was moved from `pkg/oauth`, so I already owned it. (2) It's pretty small and (3) I don't expect it to change often.

johanstokking reviewed (approved):
Please let me know if you want me to review something specifically, and if so, what

kschiffer commented:
> Please let me know if you want me to review something specifically, and if so, what

I don't think there's anything else for you to review in this.

bafonins commented on pkg/webui/account/store/middleware/user.js:
```diff
@@ -12,28 +12,37 @@
+      meta: { selector },
+    } = action
+    const userId =
+      'payload' in action && action.payload.id ? action.payload.id : selectUserId(getState())
```
it seems that id` is required here. any use case when it can be missing?

kschiffer replied:
Once the user is logged in, we can omit the user id and get it from the store. This does not work initially during the init logic. Since after the initialization the user is stored, there should be no situation where the user id is missing.

bafonins commented on pkg/webui/account/store/reducers/identity-server.js:
```diff
@@ -0,0 +1,31 @@
+
+import { handleActions } from 'redux-actions'
+
+import { GET_IS_CONFIGURATION_SUCCESS } from '@console/store/actions/identity-server'
```
I dont think we should reference console and account apps. Some repetition wont hurt until we implement some proper api to share store related functionality between apps.

bafonins replied:
Actually, there are is actions defined in `@account/store/actions/identity-server`

kschiffer replied:
Yup. Forgot to rename the import.

bafonins commented on pkg/webui/account/views/create-account/index.js:
```diff
@@ -91,151 +87,124 @@ const getSuccessMessage = state => {
-      setSubmitting(false)
-    }
+  const [error, setError] = useState(undefined)
+  const passwordRequirements = useSelector(selectPasswordRequirements)
```
You can get the is configuration directly from `useRequest` on L87

kschiffer replied:
I can but it would need an extra selector to extract the password requirements from the is config object (and not from the root of the store). I don't think it's worth the hassle.

bafonins commented on pkg/webui/account/views/login/index.js:
```diff
@@ -59,112 +57,107 @@ const validationSchema = Yup.object().shape({
 
-  return next
+  return (
+    <React.Fragment>
```
fragment not needed

bafonins commented on pkg/webui/account/views/login/index.js:
```diff
@@ -59,112 +57,107 @@ const validationSchema = Yup.object().shape({
+              to={`/register${location.search}`}
+              secondary
+              message={m.createAccount}
+              className={style.registerButton}
```
`register-button` class does not exist

bafonins commented on pkg/webui/account/views/front/index.js:
```diff
@@ -27,9 +29,12 @@ import FrontNotFound from '@account/views/front-not-found'
+const FrontView = ({ location }) => {
   return (
     <div className={style.container}>
       <section className={style.content}>
```
On the next line `.header` class does not exist.

bafonins commented on pkg/webui/console/components/user-data-form/index.js:
```diff
@@ -79,53 +74,14 @@ const validationSchema = Yup.object().shape({
     this.validationSchema = update
-      ? validationSchema
+      ? baseValidationSchema
       : createPasswordValidationSchema(passwordRequirements)
```
We need to merge password validation with validation rules for other fields as well. With this implementation:
1. Form values are no properly casted
2. Multiple fields are missing validation

bafonins commented on pkg/webui/lib/components/full-view-error/error.js:
```diff
@@ -33,10 +32,13 @@ import {
-const FullViewErrorInner = function({ error, env }) {
+const appRoot = selectApplicationRootPath()
+
+const FullViewErrorInner = function({ error }) {
```
```suggestion
const FullViewErrorInner = ({ error }) => {
```

bafonins commented on pkg/webui/lib/components/full-view-error/error.js:
```diff
@@ -88,22 +90,19 @@ const FullViewErrorInner = function({ error, env }) {
 
-const FullViewErrorInnerWithEnv = withEnv(FullViewErrorInner)
-
 const FullViewError = function({ error, header }) {
```
```suggestion
const FullViewError = ({ error, header }) => {
```

bafonins commented on pkg/webui/account.js:
```diff
@@ -15,40 +15,35 @@
   const App = require('./account/views/app').default
 
   DOM.render(
-    <EnvProvider env={env}>
```
We still have base components that rely on `withEnv` hoc, e.g. `<BaseAnchorLink />` and `<DocLink />` that can be used in the account app. We should keep the `<EnvProvider />` as long as it is used in `@ttn-lw/conponents/`.

kschiffer replied:
Searching in `pkg/webui/components` reveals that `<BaseAnchorLink />` and `<DocLink />` seem to be the only one's that still use `withEnv`. I'll add a commit to replace this with env selectors.

kschiffer committed:
console,account: Move online status store logic to global util

kschiffer committed:
console,account: Allow application specific online/offline messages

kschiffer committed:
account: Add online status logic to account app

kschiffer committed:
account: Add toast container to app scaffold

kschiffer committed:
console,account: Remove env decorator from link component

kschiffer commented:
@bafonins thanks for the review. I've addressed all your comments and also did another rebase on latest `v3.11`. I had to add [some additional commits](https://github.com/TheThingsNetwork/lorawan-stack/pull/3613/commits/a817d1de1bc44cc67d668b89ece9f5b91db04f18) to properly merge online status functionality, which was introduced via #3629.

bafonins reviewed (approved):
Manually went and tested through various pages of the account app. Didnt review the code completely, only changes since the last review.
